src/doc/rustc/src/platform-support/wasm32-unknown-emscripten.md

```diff
@@ -0,0 +1,98 @@
+# `wasm32-unknown-emscipten`
+
+**Tier: 2**
+
+The `wasm32-unknown-emscripten` target is a WebAssembly compilation target which
+uses the Emscripten compiler toolchain. Unlike the
+[`wasm32-unknown-unknown`](./wasm32-unknown-unknown.md) target, which only has
+partial support for the Rust standard library, `wasm32-unknown-emscripten`
+integrates with JavaScipt on the web to provide a more featureful standard library.
+This target is often used when compiling for the web, accessing web APIs, or
+interoperating with C and C++ code.
```

Comment:
Should we mention something about preferring `wasm32-unknown-unknown` or the `wasi` targets if C/C++ interop isn't needed? I'm not sure what kind of usecases would encourage using one over the other, but this would be great to document to steer people to the right thing.

Cc @alexcrichton

Reply:
I've added two paragraphs at the end of the section but would also appreciate Alex's input.

Reply:
I might recommend rephrasing/reorganizing this section along the lines of:

1. What is Emscripten and how this target integrates? (e.g. Emscripten is a C/C++ toolchain not related to Rust used for running C/C++ code on the web)
2. Why Emscripten is used and how Rust integrates. (e.g. Emscripten implements POSIX in terms of web APIs and provides a C standard library for that)
3. How this contrasts to other wasm targets in Rust (e.g. unlike `wasm32-unknown-unknown` it can run C/C++ and provides a full implementation of libc/libstd, unlike `wasm32-wasip1` it requires JS glue code to run on the web because the interface between wasm/JS isn't standard)

In terms of C/C++ I'd recommend focusing on both that and library support for Emscripten. Emscripten isn't just about C/C++ but also all the polyfills provided for libraries like SDL (IIRC). The contrast to `wasm32-unknown-unknown` can indicate that the unknown target has no support for C/C++ or anything that requires JS (or at least not inherently).

When comparing to `wasm32-wasip1` I'd recommend mostly contrasting it in terms of web support. The WASI target has no standard way of accessing web APIs where Emscripten has the ability to run arbitrary JS from wasm and access many web APIs. In the end the difference here will likely boil down to "are you running on the web or not" but that doesn't alone accurately capture the nuance of you can run WASI on the web but you just won't have access to the full suite of web APIs by default.

---

That may be more words than desired, sorry about that! Overall I think the wording is fine as-is, so what I'm thinking here is just minor updates here and there. Feel free to take or leave any suggestions as you see fit!

Reply:
Thank you for your extensive feedback @alexcrichton, I really appreciate it!

I've tried to address it in the latest commit, I hope I managed to capture your feedback